### ISSUE TITLE
topics to cover on git/github for week 2 lecture

### ISSUE THREAD

**Issue by ozanj** [Rucla-ed/rclass2#15]
what topics should I cover on git/github in week 2 lecture?

so far, I will have done this:

- what/why git and github
- conceptual overview of git/github flow
- command line
- basic shell commands
- basic git processes using command line [init, clone, add, commit, push, pull, etc.]

what would be next substantive topics to cover. starting in week 2, want to cover a small number of topics and go deep. so what would be next? branches/merging? deeper understanding of commits and snapshots? how to examine differences when merging? something else? @cyouh95 you are the expert here.

**Comment by cyouh95** [comment]
If Weeks 2 & 3 will be on git/Github, maybe the below 2 topics can be covered?

- Backtracking/resetting
  - `git diff <filename>`: view changes you've made to a tracked file
  - `git checkout <filename>`: discard changes you've made to file
  - `git reset HEAD <filename>`: unstage a file you've added to staging area
  - `git log`: check past commits
    - May be good to note that, if the list of past commits is long, you'll be able to use your up/down arrow keys to scroll through them, but then need to hit `q` on your keyboard to exit viewing mode
  - `git reset <commit_id>`: undo commits up to a previous commit, specified by the first 7 digits of its ID (seen from `git log`)
    - But I think this is only for if you haven't pushed those commits you want to undo to remote. Otherwise, it should probably be `git revert` to keep all commits as is and create a new commit that reverts everything to a previous commit of your choice [[x](https://www.pixelstech.net/article/1549115148-git-reset-vs-git-revert)]
- Branching/pull requests
  - `git branch`: lists branches
  - `git branch <branch_name>`: creates new branch
  - `git checkout <branch_name>`: switches to branch
  - `git merge <branch_name>`: merges specified branch to current branch
    - Or maybe use Pull Requests on Github
  - `git branch -d <branch_name>`: deletes branch

**Comment by ozanj** [comment]
@mpatricia01 @cyouh95 

These look great @cyouh95 ! can you make a recommendation about which things to cover in which order? and what should be focus of lecture 2 vs. lecture 3.

I'll use your recommendations as marching orders for me. cuz I don't really know these topics myself so I'll read/learn them first and then start writing lecture text around them.

**Comment by cyouh95** [comment]
Maybe things under "Backtracking/resetting" for Lecture 2 and "Branching/pull requests" for Lecture 3?

**Comment by ozanj** [comment]
Cool. Would merging go with branching?

**Comment by cyouh95** [comment]
Yup! Branching and merging are both under "Branching/pull requests" above

**Comment by ozanj** [comment]
👍

**Comment by ozanj** [comment]
@cyouh95 

Can you try to find some tutorial type readings about the topics you listed above. these readings would be basis of me learning these things. and readings that I could potentially assign to students.
